crates/typst/src/visualize/pattern.rs

```diff
@@ -0,0 +1,241 @@
+///   [`rotate`]($rotate) will not affect the parent of a gradient, but a
+///   [`grid`]($grid) will.
+#[ty(scope)]
+#[derive(Debug, Clone, PartialEq, Hash)]
+pub struct Pattern {
+    /// The body of the pattern
+    pub body: Prehashed<Content>,
+    /// The pattern's rendered content.
+    pub frame: Prehashed<Frame>,
+    /// The pattern's tile size.
+    pub bbox: Size,
+    /// The pattern's tile spacing.
+    pub spacing: Size,
+    /// The pattern's relative transform.
+    pub relative: Smart<RelativeTo>,
+}
+
+impl Eq for Pattern {}
+
+#[scope]
+impl Pattern {
+    /// Construct a new pattern.
+    ///
+    /// ```example
+    /// #let pat = pattern(
+    ///   (20pt, 20pt),
+    ///   relative: "parent",
+    ///   place(dx: 5pt, dy: 5pt, rotate(45deg, square(size: 5pt, fill: black)))
+    /// );
+    ///  #rect(width: 100%, height: 100%, fill: pat)
+    /// ```
+    #[func(constructor)]
+    pub fn construct(
+        vm: &mut Vm,
+        /// The bounding box of each cell of the pattern.
+        bbox: Spanned<Axes<Length>>,
+        /// The content of each cell of the pattern.
+        body: Content,
+        /// The spacing between cells of the pattern.
+        #[named]
+        #[default(Spanned::new(Axes::splat(Length::zero()), Span::detached()))]
+        spacing: Spanned<Axes<Length>>,
+        /// The [relative placement](#relativeness) of the pattern.
+        ///
+        /// For an element placed at the root/top level of the document, the
+        /// parent is the page itself. For other elements, the parent is the
+        /// innermost block, box, column, grid, or stack that contains the
+        /// element.
+        #[named]
+        #[default(Smart::Auto)]
+        relative: Smart<RelativeTo>,
+    ) -> SourceResult<Pattern> {
+        // Ensure that sizes are absolute.
+        if !bbox.v.x.em.is_zero() || !bbox.v.y.em.is_zero() {
+            bail!(bbox.span, "pattern tile size must be absolute");
+        }
+
+        // Ensure that sizes are non-zero and finite.
+        if bbox.v.x.is_zero()
+            || bbox.v.y.is_zero()
+            || !bbox.v.x.is_finite()
+            || !bbox.v.y.is_finite()
+        {
+            bail!(bbox.span, "pattern tile size must be non-zero and non-infinite");
+        }
+
+        // Ensure that spacing is absolute.
+        if !spacing.v.x.em.is_zero() || !spacing.v.y.em.is_zero() {
+            bail!(spacing.span, "pattern tile spacing must be absolute");
+        }
+
+        // Ensure that spacing is finite.
+        if !spacing.v.x.is_finite() || !spacing.v.y.is_finite() {
+            bail!(spacing.span, "pattern tile spacing must be finite");
+        }
+
+        // The size of the pattern.
+        let size = Size::new(bbox.v.x.abs, bbox.v.y.abs);
+
+        // Layout the pattern.
+        let regions = Regions::one(size, Axes::splat(false));
+        let mut frame =
+            body.layout(&mut vm.vt, StyleChain::default(), regions)?.into_frame();
```

Comment:
```suggestion
        let library = world.library();
        let styles = StyleChain::new(&library.styles);
        let regions = Regions::one(size, Axes::splat(false));
        let mut frame = body.layout(&mut vm.vt, styles, regions)?.into_frame();
```
we should probably respect at least the standard library styles (i.e. also those of the test runner).

this will also not allow for introspection, but that's not really a problem probably. with styles I'm less sure (I know we talked about this before).